### PR TITLE
review: refactor: CtTypeParameterReference

### PR DESCRIPTION
**Pull request by pvojtechovsky**
an experiment related to #2401

### CODE REVIEW

**Comment by pvojtechovsky** [comment]
All the tests should pass now - I just did a rebase to minimize amount of changes in this PR.

This PR implements ideas from #2401 and fixes #2161 (the failing test is already part of this PR) and probably fixes also #2378 (I have not added that test here).

The main conceptual changes are:
* CtTypeParameterReference has no bounds and no super class. It holds only `simpleName` of type parameter - the methods were moved from CtTypeParameterReference to CtWildcardReference and some of them were even removed, because CtWildcardReference doesn't support `CtIntersectionTypeReference`. So both types now mirrors java features.
* CtExecutableReference#parameters must not contain `CtTypeParameterReference` (it will fail with SpoonException now), but they must contain their bounds/superClass instead.

Q1) WDYT?

I suggest another change (which is not part of this PR): `CtWildcardReference extends CtTypeReference<Object>` ... actually it extends `CtTypeParameterReference` and it seems to me confusing and makes check
```java
ref instanceof CtTypeParameterReference` 
```
tricky, because it usually have to look like this
```java
ref instanceof CtTypeParameterReference && !(ref instanceof CtWildcardReference)
```

Q2) WDYT?

And I am not sure what should we use for return type of `CtExectutableReference`
A) CtTypeParameterReference - same like before - current implementation of this PR
B) CtTypeReference with bounds/superClass of `CtTypeParameterReference` - same like CtExectutableReference parameters.

Q3) WDYT?

File: src/test/java/spoon/test/parent/ParentTest.java [hunk 1]
```diff
@@ -212,10 +212,10 @@ public void testParentOfPrimitiveReference() throws Exception {
 	}
 
 	public static void checkParentContract(CtPackage pack) {
-		for(CtElement elem: pack.getElements(new TypeFilter<>(CtElement.class))) {
+		pack.filterChildren(null).forEach((CtElement elem) -> {
 			// there is always one parent
-			Assert.assertNotNull("no parent for "+elem.getClass()+"-"+elem.getPosition(), elem.getParent());
```
Comment (monperrus) [inline src/test/java/spoon/test/parent/ParentTest.java]: AFAIU we can have both assertions there

**Comment by monperrus** [comment]
Q1) It looks good! There are very few changes in the tests, which indeed shows that this was unnecessary

Q2) Maybe, but it seems we can discuss this separately in another PR #babypr 

Q3) Return type of which method?

**Comment by pvojtechovsky** [comment]
> Q3) Return type of which method?

```java
interface A<T extends String> {
 T method();
}
```
and now there is question what `CtExecutableReference.getType()` should return:
A) CtTypeParameterReference -> `T`
B) CtTypeReference -> `String`

**Comment by monperrus** [comment]
I would say we keep the current behavior for backward compatibility and we add a method "getResolvedType" or "getInferredType" with the new behavior.

**Comment by monperrus** [comment]
I guess that we're ready for merge?

**Comment by pvojtechovsky** [comment]
Yes this PR is ready for merge

**Comment by monperrus** [comment]
Thanks Pavel for this conceptual cleaning

**Pull request merged**